### DISCUSSION TITLE
findMany together with additional aggregation

### DISCUSSION THREAD

**Discussion by anton-johansson** [prisma/prisma#15874] [Q&A]
## Problem

Imagine the following model:

```
model User {
    userId                          Int                 @id @default(autoincrement())
    emailAddress                    String              @unique @db.NVarChar(100)
    inviteTokens                    InviteToken[]
}

model InviteToken {
    tokenId                         Int                 @id @default(autoincrement())
    token                           String              @unique @db.VarChar(32)
    userId                          Int
    user                            User                @relation(fields: [userId], references: [userId])
    expiresAt                       DateTime
}
```

Now, imagine that I want to fetch all users and their latest invitation token. With raw SQL, I would write this:

```sql
SELECT  [user].userId
,       [user].emailAddress
,       inviteToken.expiresAt
FROM	[User] [user]
OUTER APPLY
(
        SELECT  MAX(token.expiresAt) [expiresAt]
        FROM    InviteToken token
        WHERE   token.userId = instance.userId
) [inviteToken]
ORDER BY [user].emailAddress
```

So this is my current query, but it feels very inconvenient to have a separate query and have to manually map together the aggregated values to the results of the first query.

```typescript
const users = await client.user.findMany({
    select: {
        userId: true,
        emailAddress: true,
    },
    orderBy: {
        emailAddress: "asc",
    },
});

const expiresAt = await client.inviteToken.groupBy({
    by: ["userId"],
    _max: {
        expiresAt: true,
    },
    where: {
        userId: {
            in: users.map(user => user.userId),
        },
    },
});
```

## Suggested solution

Not sure... Any ideas?

## Alternatives

None that I've found.

## Additional context

I'm using Microsoft SQL server and latest Prisma (4.4.0).

**Comment by floelhoeffel** [comment]
Would this work?
```ts
const users = await prisma.user.findMany({
  select: {
    userId: true,
    emailAddress: true,
    inviteTokens: {
      select: {
        expiresAt: true,
      },
      orderBy: [{ expiresAt: "desc" }],
      take: 1,
    },
  },
});
```

**Reply by vladi-stevanovic** [reply]
:wave: Hello @anton-johansson , I’m going through pending GitHub Discussions to check if you need any help. Is this issue still open?

Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! :raised_hands:

**Reply by anton-johansson** [reply]
Hey @floelhoeffel and @vladi-stevanovic, sorry for the late reply. Not sure how I missed this.

The above query could work, I'll try to refactor my code and see if it does.

But what about other types of aggregations? For example, what if I wanted to count something?

Like this (raw SQL):
```sql
SELECT  type.campaignTypeId
,       type.campaignTypeName
,       x.count
FROM    CampaignType type
OUTER APPLY
(
        SELECT  COUNT(1) [count]
        FROM    Campaign campaign
        WHERE   campaign.campaignTypeId = type.campaignTypeId
        AND     campaign.active = 1
) x
ORDER BY type.campaignTypeId
```

I'm looking at something like this:
```ts
const types = await client.campaignType.findMany({
    select: {
        campaignTypeId: true,
        campaignTypeName: true,
        _count: {
            select: {
                campaigns: true,
            },
        },
    },
    orderBy: {
        campaignTypeId: "asc",
    },
});
```

But it does not allow me to perform additional `WHERE`-statements, like `campaign.active = 1`. Any ideas?

**Reply by anton-johansson** [reply]
Similar to my original post, this is what I'm doing now. It works, but is far from optimal. :smile: 

```ts
const types = await client.campaignType.findMany({
    select: {
        campaignTypeId: true,
        campaignTypeName: true,
    },
    orderBy: {
        campaignTypeId: "asc",
    },
});


// Workaround for discussion:
// https://github.com/prisma/prisma/discussions/15874
const count = await client.campaign.groupBy({
    by: ["campaignTypeId"],
    _count: {
        campaignTypeId: true,
    },
    where: {
        active: true,
    },
});

return types.map(type => ({
    ...type,
    numberOfCampaigns: count.find(data => data.campaignTypeId === hit.campaignTypeId)?._count.campaignTypeId ?? 0, // See comment about workaround above
}));
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
